language/testing-infra/transactional-test-runner/src/tasks.rs

```diff
@@ -0,0 +1,319 @@
+        let Self {
+            command,
+            name,
+            number,
+            start_line,
+            command_lines_stop,
+            stop_line,
+            data,
+        } = self;
+        match split(command) {
+            Either::Left(cmd) => Either::Left(TaskInput {
+                command: cmd,
+                name,
+                number,
+                start_line,
+                command_lines_stop,
+                stop_line,
+                data,
+            }),
+            Either::Right(subcmd) => Either::Right(TaskInput {
+                command: subcmd,
+                name,
+                number,
+                start_line,
+                command_lines_stop,
+                stop_line,
+                data,
+            }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SyntaxChoice {
+    Source,
+    IR,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum ProvidedCommand {
+    #[structopt(name = "publish")]
+    Publish {
+        #[structopt(long = "gas-budget")]
+        gas_budget: Option<u64>,
+        #[structopt(long = "syntax")]
+        syntax: Option<SyntaxChoice>,
+        #[structopt(long = "address", parse(try_from_str = parse_account_address))]
+        address: Option<AccountAddress>,
+    },
+
+    #[structopt(name = "run")]
+    Run {
+        #[structopt(long = "signers", parse(try_from_str = parse_account_address))]
+        signers: Vec<AccountAddress>,
+        #[structopt(long = "args", parse(try_from_str = parser::parse_transaction_argument))]
+        args: Vec<TransactionArgument>,
+        #[structopt(long = "type-args", parse(try_from_str = parser::parse_type_tag))]
+        type_args: Vec<TypeTag>,
+        #[structopt(long = "gas-budget")]
+        gas_budget: Option<u64>,
+        #[structopt(long = "syntax")]
+        syntax: Option<SyntaxChoice>,
+        #[structopt(name = "NAME", parse(try_from_str = parse_qualified_module_access))]
+        name: Option<(ModuleId, Identifier)>,
+    },
+
+    #[structopt(name = "view")]
+    View {
+        #[structopt(long = "address", parse(try_from_str = parse_account_address))]
+        address: AccountAddress,
+        #[structopt(long = "resource", parse(try_from_str = parse_qualified_module_access_with_type_args))]
+        resource: (ModuleId, Identifier, Vec<TypeTag>),
+    },
+}
+
+#[macro_export]
+macro_rules! define_commands {
+    ($command:ident $(, $subcommand:ident)?) => {
+        #[allow(unused_imports)]
+        use structopt::*;
+        #[derive(Debug, StructOpt)]
+        #[structopt(
+            name = "task",
+            about = "Transactional task commands",
+            rename_all = "kebab-case"
+        )]
+        pub enum $command {
+            #[structopt(flatten)]
+            Normal($crate::tasks::ProvidedCommand),
+
+            $(
+                #[structopt(flatten)]
+                Subcommand($subcommand),
+            )?
+        }
+    }
+
+}
```

Comment:
Could I ask why we are defining this macro instead of using generics? Is it because of structopt? 
```
enum Command<S> {
    Provided(ProvidedCommand),
    Subcommand(S),
}
```

From a trait implementor's point of view, it is very confusing that I need to define an enum of (ProvidedCommand + Subcommand) and implement a split function, given that I've already specified the associated Subcommand type. IMO this should be done by the framework.

Reply:
StructOpt can't handle generics so we are stuck doing some wonky stuff

Reply:
It's also quite confusing to see that for the vm test harness,`EmptySubcommand` is not part of the enum definition. I mean, I get it, it's an empty command and we don't really need to process any instance of it, but it still bothers me that we are defining the associated type, but then see that there's only one enum variant to check, and yet the match is exhaustive:
```
run_test_impl(path, &mut adaptor, |c| match c {
    TaskCommand::Normal(p) => Either::Left(p),
})
```

Reply:
> StructOpt can't handle generics so we are stuck doing some wonky stuff

Let me see if it's possible to implement the trait manually without automatic derivation.

Reply:
> I still think that if possible, we shouldn't require the client to implement a split function, which is a leaky abstraction.

I don't worry too too much about leaky abstractions here. We are building a very simple testing framework. Things can be a tad bit leaky